include/zephyr/drivers/input.h

```diff
@@ -0,0 +1,307 @@
+	 * Custom data size must be less than or equal to CONFIG_INPUT_ATTR_DATA_SIZE.
+	 */
+	uint32_t userdata[CONFIG_INPUT_ATTR_DATA_SIZE / sizeof(uint32_t)];
+};
+
+BUILD_ASSERT(sizeof(union input_attr_data) <= CONFIG_INPUT_ATTR_DATA_SIZE);
+
+/**
+ * @brief Representation of a input event.
+ *
+ */
+struct input_event {
+	/* The time of the event, the time is the system uptime */
+	struct timeval time;
+	/* event type (EV_SYN, EV_KEY, EV_ABS, etc) */
+	uint16_t type;
+	/* event code (Keyboard code, ABS code, etc) */
+	uint16_t code;
+	/* event value (Keyboard value, ABS position, etc) */
+	int32_t value;
+};
+
+#define input_event_size	sizeof(struct input_event)
+
+/**
+ * @brief Event type
+ *
+ */
+#define EV_SYN			0x00
+#define EV_KEY			0x01
+#define EV_REL			0x02
+#define EV_ABS			0x03
+
+/**
+ * @brief Synchronization events
+ *
+ */
+#define SYN_REPORT		0x00
+
+/**
+ * @brief Relative axes
+ *
+ */
+#define REL_X			0x00
+#define REL_Y			0x01
+#define REL_Z			0x02
+
+/**
+ * @brief Absolute axes
+ *
+ */
+#define ABS_X			0x00
+#define ABS_Y			0x01
+#define ABS_Z			0x02
+
+/**
+ * @brief Event Key value
+ *
+ */
+enum keyboard_value {
+	KEY_RELEASE,
+	KEY_PRESSED,
+	KEY_LONG_PRESSED,
+	KEY_HOLD_PRESSED,
+	KEY_LONG_RELEASE
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * Input subsystem driver API definition and system call entry points.
+ *
+ * (Internal use only.)
+ */
+typedef int (*input_setup_t)(const struct device *dev);
+typedef int (*input_release_t)(const struct device *dev);
+typedef int (*input_attr_get_t)(const struct device *dev,
+			enum input_attr_type type, union input_attr_data *data);
+typedef int (*input_attr_set_t)(const struct device *dev,
+			enum input_attr_type type, union input_attr_data *data);
+typedef int (*input_event_read_t)(const struct device *dev, struct input_event *event);
+typedef int (*input_event_write_t)(const struct device *dev, struct input_event *event);
+
+__subsystem struct input_driver_api {
+	input_setup_t setup;
+	input_release_t release;
+	input_attr_get_t attr_get;
+	input_attr_set_t attr_set;
+	input_event_read_t event_read;
+	input_event_write_t event_write;
+};
+/**
+ * @endcond
+ */
+
+/**
+ * @brief Setup a input device instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int input_setup(const struct device *dev);
+
+static inline int z_impl_input_setup(const struct device *dev)
+{
+	const struct input_driver_api *api =
+				(struct input_driver_api *)dev->api;
+
+	return api->setup(dev);
+}
+
+/**
+ * @brief Release a input device instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int input_release(const struct device *dev);
+
+static inline int z_impl_input_release(const struct device *dev)
```

Comment:
what would be the use case for this function?

Like the setup above. It seems to bring concepts (pluggin in/out a keyboard? pm perhaps?) to the user level that should not be its role to manage. Which is not the right way.